File: app/src/main/java/com/aar/app/wsp/features/gamethemeselector/ThemeSelectorActivity.java
package com.aar.app.wsp.features.gamethemeselector;

import android.annotation.SuppressLint;
import androidx.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.aar.app.wsp.R;
import com.aar.app.wsp.WordSearchApp;
import com.aar.app.wsp.custom.easyadapter.MultiTypeAdapter;
import com.aar.app.wsp.features.FullscreenActivity;
import com.aar.app.wsp.features.ViewModelFactory;
import com.aar.app.wsp.model.GameTheme;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import io.reactivex.disposables.Disposable;

public class ThemeSelectorActivity extends FullscreenActivity {

    public static final String EXTRA_THEME_ID = "game_theme_id";
    public static final String EXTRA_ROW_COUNT = "row_count";
    public static final String EXTRA_COL_COUNT = "col_count";

    @BindView(R.id.loadingLayout) ViewGroup mLoadingLayout;
    @BindView(R.id.progressBar) ProgressBar mProgressBar;
    @BindView(R.id.rvThemes) RecyclerView mRvThemes;
    @BindView(R.id.textRev) TextView mTextRev;
    @BindView(R.id.btnUpdate) Button mBtnUpdate;

    @Inject
    ViewModelFactory mViewModelFactory;
    ThemeSelectorViewModel mViewModel;

    private Disposable mUpdateDisposable;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_theme_selector);

        ((WordSearchApp) getApplication()).getAppComponent().inject(this);
        ButterKnife.bind(this);

        MultiTypeAdapter adapter = new MultiTypeAdapter();
        adapter.addDelegate(
                GameThemeItem.class,
                R.layout.item_theme_list,
                (model, holder) -> {
                    holder.<TextView>find(R.id.textTheme).setText(model.getName());
                    holder.<TextView>find(R.id.textCount).setText(
                            getString(R.string.text_words)
                                    .replaceAll(":count", String.valueOf(model.getWordsCount()))
                    );
                },
                (model, view) -> onItemClick(model.getId())
        );

        mRvThemes.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false));
        mRvThemes.addItemDecoration(new DividerItemDecoration(this, LinearLayoutManager.VERTICAL));
        mRvThemes.setAdapter(adapter);

        mViewModel = ViewModelProviders.of(this, mViewModelFactory).get(ThemeSelectorViewModel.class);
        mViewModel.getOnGameThemeLoaded().observe(this, gameThemes -> {
            adapter.setItems(gameThemes);
            mRvThemes.setVisibility(View.VISIBLE);
            mProgressBar.setVisibility(View.GONE);
        });
        updateRevisionNumber();

        loadData();
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (mUpdateDisposable != null) {
            mUpdateDisposable.dispose();
        }
    }

    @OnClick(R.id.btnAllTheme)
    public void onAllThemeClick() {
        onItemClick(GameTheme.NONE.getId());
    }

    @OnClick(R.id.btnUpdate)
    public void onUpdateClick() {
        mLoadingLayout.setVisibility(View.VISIBLE);
        mBtnUpdate.setEnabled(false);
        mUpdateDisposable = mViewModel.updateData()
                .subscribe(responseType -> {
                            updateRevisionNumber();
                            mLoadingLayout.setVisibility(View.GONE);
                            mBtnUpdate.setEnabled(true);
                            String message;
                            if (responseType == ThemeSelectorViewModel.ResponseType.NoUpdate) {
                                message = getString(R.string.up_to_date);
                            } else {
                                message = getString(R.string.update_success);
                            }
                            Toast.makeText(
                                    ThemeSelectorActivity.this,
                                    message,
                                    Toast.LENGTH_LONG)
                                    .show();
                        }, throwable -> {
                            mLoadingLayout.setVisibility(View.GONE);
                            mBtnUpdate.setEnabled(true);
                            Toast.makeText(
                                    ThemeSelectorActivity.this,
                                    R.string.err_no_connect,
                                    Toast.LENGTH_LONG)
                                    .show();
                        }
                );
    }

    private void loadData() {
        mRvThemes.setVisibility(View.GONE);
        mProgressBar.setVisibility(View.VISIBLE);
        mViewModel.loadThemes();
    }

    @SuppressLint("CheckResult")
    private void onItemClick(int themeId) {
        mViewModel
                .checkWordAvailability(
                        themeId,
                        getGridRowCount(),
                        getGridColCount())
                .subscribe(available -> {
                    if (available) {
                        Intent intent = new Intent();
                        intent.putExtra(EXTRA_THEME_ID, themeId);
                        setResult(RESULT_OK, intent);
                        finish();
                    } else {
                        Toast.makeText(
                                ThemeSelectorActivity.this,
                                "No words data to use, please select another theme or change grid size",
                                Toast.LENGTH_LONG).show();
                    }
                });
    }

    private void updateRevisionNumber() {
        String rev = "-";
        if (mViewModel.getLastDataRevision() > 0)
            rev = String.valueOf(mViewModel.getLastDataRevision());
        mTextRev.setText(rev);
    }

    private int getGridRowCount() {
        Bundle extras = getIntent().getExtras();
        if (extras != null) {
            return extras.getInt(EXTRA_ROW_COUNT);
        }
        return 0;
    }

    private int getGridColCount() {
        Bundle extras = getIntent().getExtras();
        if (extras != null) {
            return extras.getInt(EXTRA_COL_COUNT);
        }
        return 0;
    }
}

File: app/src/main/java/com/aar/app/wsp/features/gamethemeselector/ThemeSelectorActivity.kt
package com.aar.app.wsp.features.gamethemeselector

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.lifecycle.observe
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.aar.app.wsp.R
import com.aar.app.wsp.WordSearchApp
import com.aar.app.wsp.commons.gone
import com.aar.app.wsp.commons.visible
import com.aar.app.wsp.custom.easyadapter.MultiTypeAdapter
import com.aar.app.wsp.custom.easyadapter.SimpleAdapterDelegate
import com.aar.app.wsp.features.FullscreenActivity
import com.aar.app.wsp.features.gamethemeselector.ThemeSelectorViewModel.ResponseType
import com.aar.app.wsp.model.GameTheme
import io.reactivex.disposables.Disposable
import kotlinx.android.synthetic.main.activity_theme_selector.*
import javax.inject.Inject

class ThemeSelectorActivity : FullscreenActivity() {
    @JvmField
    @Inject
    var mViewModelFactory: ViewModelProvider.Factory? = null
    lateinit var viewModel: ThemeSelectorViewModel

    private val adapter = MultiTypeAdapter()

    private var mUpdateDisposable: Disposable? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_theme_selector)
        (application as WordSearchApp).appComponent.inject(this)

        initViews()
        initRecyclerView()
        initViewModel()

        updateRevisionNumber()
        loadData()
    }

    private fun initViews() {
        btnAllTheme.setOnClickListener {
            onItemClick(GameTheme.NONE.id)
        }
        btnUpdate.setOnClickListener {
            onUpdateClick()
        }
    }

    private fun initViewModel() {
        viewModel = ViewModelProviders.of(this, mViewModelFactory).get(ThemeSelectorViewModel::class.java)
        viewModel.onGameThemeLoaded.observe(this) { gameThemes: List<GameThemeItem>? ->
            adapter.setItems(gameThemes)
            rvThemes.visible()
            progressBar.gone()
        }
    }

    private fun initRecyclerView() {
        adapter.addDelegate(
            GameThemeItem::class.java,
            R.layout.item_theme_list,
            object : SimpleAdapterDelegate.Binder<GameThemeItem> {
                override fun bind(model: GameThemeItem, holder: SimpleAdapterDelegate.ViewHolder?) {
                    holder?.find<TextView>(R.id.textTheme)?.text = model.name
                    holder?.find<TextView>(R.id.textCount)?.text =
                        getString(R.string.text_words)
                            .replace(":count".toRegex(), java.lang.String.valueOf(model.wordsCount))
                }
            },
            object : SimpleAdapterDelegate.OnItemClickListener<GameThemeItem> {
                override fun onClick(model: GameThemeItem, view: View?) {
                    onItemClick(model.id)
                }
            }
        )

        rvThemes.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        rvThemes.addItemDecoration(DividerItemDecoration(this, LinearLayoutManager.VERTICAL))
        rvThemes.adapter = adapter
    }

    override fun onStop() {
        super.onStop()
        mUpdateDisposable?.dispose()
    }

    private fun onUpdateClick() {
        loadingLayout.visibility = View.VISIBLE
        btnUpdate.isEnabled = false
        mUpdateDisposable = viewModel.updateData()
            .subscribe({ responseType: ResponseType ->
                updateRevisionNumber()
                loadingLayout.visibility = View.GONE
                btnUpdate.isEnabled = true
                val message = if (responseType == ResponseType.NoUpdate) {
                    getString(R.string.up_to_date)
                } else {
                    getString(R.string.update_success)
                }
                Toast.makeText(
                    this@ThemeSelectorActivity,
                    message,
                    Toast.LENGTH_LONG)
                    .show()
            }
            ) {
                loadingLayout.visibility = View.GONE
                btnUpdate.isEnabled = true
                Toast.makeText(
                    this@ThemeSelectorActivity,
                    R.string.err_no_connect,
                    Toast.LENGTH_LONG)
                    .show()
            }
    }

    private fun loadData() {
        rvThemes.gone()
        progressBar.visible()
        viewModel.loadThemes()
    }

    @SuppressLint("CheckResult")
    private fun onItemClick(themeId: Int) {
        viewModel
            .checkWordAvailability(
                themeId,
                gridRowCount,
                gridColCount)
            .subscribe { available: Boolean ->
                if (available) {
                    val intent = Intent()
                    intent.putExtra(EXTRA_THEME_ID, themeId)
                    setResult(Activity.RESULT_OK, intent)
                    finish()
                } else {
                    Toast.makeText(
                        this@ThemeSelectorActivity,
                        "No words data to use, please select another theme or change grid size",
                        Toast.LENGTH_LONG).show()
                }
            }
    }

    private fun updateRevisionNumber() {
        textRev.text =
            if (viewModel.lastDataRevision > 0) viewModel.lastDataRevision.toString()
            else "-"
    }

    private val gridRowCount: Int
        get() {
            val extras = intent.extras
            return extras?.getInt(EXTRA_ROW_COUNT) ?: 0
        }

    private val gridColCount: Int
        get() {
            val extras = intent.extras
            return extras?.getInt(EXTRA_COL_COUNT) ?: 0
        }

    companion object {
        const val EXTRA_THEME_ID = "game_theme_id"
        const val EXTRA_ROW_COUNT = "row_count"
        const val EXTRA_COL_COUNT = "col_count"
    }
}